File: harness/compare_c3d_c4d.c
// compare_c3d_c4d — the validation backbone (PLAN §0.5). Runs the REFERENCE
// codec c3d and OUR configured codec on the SAME raw u8 volume across a q sweep,
// emitting one machine-readable row per (codec,q) with ratio / PSNR / MS-SSIM /
// GMSD / enc-MB/s / dec-MB/s. c4d is driven separately by the Python wrapper
// (it has a clean CLI); this binary covers OURS + c3d (a C library, no CLI).
//
// Build: linked against libvc (our codec) and libc3d.a (the reference). The
// CMake/standalone build passes -DHAVE_C3D when c3d is available; without it,
// only our codec is benchmarked (graceful per PLAN: "if a reference won't build,
// note it and continue").
//
// Usage: compare_c3d_c4d <raw.u8> <dz> <dy> <dx> <label> <q...>
// Output (stdout), one row per line, TSV:
//   codec<TAB>label<TAB>q<TAB>ratio<TAB>psnr<TAB>ms_ssim<TAB>gmsd<TAB>enc_mbs<TAB>dec_mbs<TAB>tag
// where `tag` identifies the active config (entropy+qweight) for our codec.
#define _POSIX_C_SOURCE 199309L   // clock_gettime / CLOCK_MONOTONIC under -std=c23
#include "../include/vc/vc.h"
#include "../src/metrics/metrics.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifdef HAVE_C3D
#include "c3d.h"
#endif

#ifndef VC_TAG_STR
#define VC_TAG_STR "ours"
#endif

static double now_sec(void) {
    struct timespec ts; clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}
static u8 *read_file(const char *path, size_t *len) {
    FILE *f = fopen(path, "rb"); if (!f) return NULL;
    fseek(f, 0, SEEK_END); long sz = ftell(f); fseek(f, 0, SEEK_SET);
    u8 *b = (u8 *)malloc(sz);
    if (fread(b, 1, sz, f) != (size_t)sz) { free(b); fclose(f); return NULL; }
    fclose(f); *len = (size_t)sz; return b;
}

static void emit(const char *codec, const char *label, double q, double ratio,
                 const u8 *ref, const u8 *rec, u32 d, u32 h, u32 w,
                 double enc_mbs, double dec_mbs, const char *tag) {
    vc_metrics m; vc_compute_metrics(ref, rec, d, h, w, &m);
    double gmsd = vc_gmsd(ref, rec, d, h, w);
    printf("%s\t%s\t%.4g\t%.2f\t%.3f\t%.4f\t%.4f\t%.1f\t%.1f\t%s\n",
           codec, label, q, ratio, m.psnr, m.ms_ssim, gmsd, enc_mbs, dec_mbs, tag);
    fflush(stdout);
}

int main(int argc, char **argv) {
    if (argc < 7) {
        fprintf(stderr, "usage: %s <raw.u8> <dz> <dy> <dx> <label> <q...>\n", argv[0]);
        return 2;
    }
    const char *path = argv[1];
    u32 d = (u32)atoi(argv[2]), h = (u32)atoi(argv[3]), w = (u32)atoi(argv[4]);
    const char *label = argv[5];
    size_t len; u8 *vol = read_file(path, &len);
    if (!vol || len < (size_t)d * h * w) { fprintf(stderr, "bad raw %s\n", path); return 1; }
    const size_t raw = (size_t)d * h * w;

    for (int qi = 6; qi < argc; ++qi) {
        double q = atof(argv[qi]);

        // --- OUR codec ---
        {
            u8 *arc = NULL; size_t alen = 0;
            double t0 = now_sec();
            if (vc_encode_volume(vol, d, h, w, (f32)q, &arc, &alen)) { fprintf(stderr, "ours enc fail\n"); continue; }
            double t1 = now_sec();
            u8 *rec = NULL; u32 rd, rh, rw;
            if (vc_decode_volume(arc, alen, &rec, &rd, &rh, &rw)) { fprintf(stderr, "ours dec fail\n"); free(arc); continue; }
            double t2 = now_sec();
            emit("ours", label, q, (double)raw / (double)alen, vol, rec, d, h, w,
                 raw / 1e6 / (t1 - t0), raw / 1e6 / (t2 - t1), VC_TAG_STR);
            free(arc); free(rec);
        }

#ifdef HAVE_C3D
        // --- c3d reference (256^3 chunk atom; encode_at_q over chunk grid) ---
        // c3d's codec atom is a fixed 256^3 chunk. We tile the volume into 256^3
        // chunks (zero-padded), encode each at q, decode, scatter back, then
        // score the full volume. NOTE: c3d's q is an INVERSE quantizer scalar
        // (small q == high quality, [2^-6, 2^12]) on a different scale than our
        // dead-zone step. We therefore sweep c3d over its OWN native q grid
        // (mapped from our q index) so its rate-distortion curve is sampled
        // across the same 5-200x band, then compare ratio-at-PSNR (not at equal
        // q). The mapping below was calibrated on PHerc Paris 4 (see report).
        {
            // map our sweep index to a c3d-native q spanning ~0.015..0.5
            static const double C3DQ[] = {0.015625, 0.03125, 0.0625, 0.125, 0.25, 0.5, 1.0};
            int idx = qi - 6; if (idx > 6) idx = 6;
            double cq = C3DQ[idx];
            const u32 C = C3D_CHUNK_SIDE;            // 256
            u32 ncz = (d + C - 1) / C, ncy = (h + C - 1) / C, ncx = (w + C - 1) / C;
            // c3d requires 32-byte-aligned 256^3 voxel buffers (C3D_ALIGN).
            u8 *cin  = (u8 *)aligned_alloc(32, (size_t)C * C * C);
            u8 *cout = (u8 *)aligned_alloc(32, (size_t)C * C * C);
            u8 *enc  = (u8 *)malloc(c3d_chunk_encode_max_size());
            u8 *rec  = (u8 *)calloc(raw, 1);
            size_t total_enc = 0;
            double enc_t = 0, dec_t = 0;
            for (u32 cz = 0; cz < ncz; ++cz)
            for (u32 cy = 0; cy < ncy; ++cy)
            for (u32 cx = 0; cx < ncx; ++cx) {
                memset(cin, 0, (size_t)C * C * C);
                for (u32 z = 0; z < C && cz*C+z < d; ++z)
                for (u32 y = 0; y < C && cy*C+y < h; ++y) {
                    u32 vz = cz*C+z, vy = cy*C+y, ox = cx*C;
                    u32 cw = (ox + C <= w) ? C : (ox < w ? w - ox : 0);
                    if (cw) memcpy(cin + ((size_t)z*C+y)*C,
                                   vol + (((size_t)vz*h+vy)*w+ox), cw);
                }
                double a = now_sec();
                size_t el = c3d_chunk_encode_at_q(cin, (float)cq, enc, c3d_chunk_encode_max_size());
                double b = now_sec();
                c3d_chunk_decode(enc, el, cout);
                double c = now_sec();
                enc_t += b - a; dec_t += c - b; total_enc += el;
                for (u32 z = 0; z < C && cz*C+z < d; ++z)
                for (u32 y = 0; y < C && cy*C+y < h; ++y) {
                    u32 vz = cz*C+z, vy = cy*C+y, ox = cx*C;
                    u32 cw = (ox + C <= w) ? C : (ox < w ? w - ox : 0);
                    if (cw) memcpy(rec + (((size_t)vz*h+vy)*w+ox),
                                   cout + ((size_t)z*C+y)*C, cw);
                }
            }
            emit("c3d", label, cq, (double)raw / (double)total_enc, vol, rec, d, h, w,
                 raw / 1e6 / enc_t, raw / 1e6 / dec_t, "ref-c3d");
            free(cin); free(cout); free(enc); free(rec);
        }
#endif
    }
    free(vol);
    return 0;
}

File: harness/compare_c3d_c4d.py
#!/usr/bin/env python3
"""compare_c3d_c4d — Phase 0.5 validation backbone (PLAN §0.5).

Runs the reference codecs c3d and c4d AND our configured codec on the SAME
PHerc Paris 4 inputs the rest of the harness uses, and emits a side-by-side
ratio / PSNR / MS-SSIM / GMSD / enc-MB/s / dec-MB/s table.

Strategy
--------
* OURS + c3d : a single linked C binary (compare_c3d_c4d.c) that runs our codec
  (one build per entropy x qweight config) and, when linked with libc3d, the
  c3d reference, on the same raw volume across a q sweep. We build the 9 our-
  codec configs (rice/rlgr/rans x flat/HF/adaptive) once each.
* c4d        : driven via its clean CLI (encode/decode), metrics computed by a
  tiny scorer that reuses our metrics (vc_score helper binary).

If a reference fails to build, we note it and continue with whatever built
(PLAN mandate). All builds are out-of-tree under harness/refbuild/.

Usage:
  compare_c3d_c4d.py [--inputs hires|coarse|both] [--q ...]
"""
import os, sys, subprocess, struct, shutil, argparse, tempfile

ROOT   = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
C3D    = os.path.join(os.path.dirname(ROOT), "c3d")
C4D    = os.path.join(os.path.dirname(ROOT), "c4d")
REFB   = os.path.join(ROOT, "harness", "refbuild")
DATA   = os.path.join(ROOT, "data")

ENTROPIES = ["rice", "rlgr", "rans"]
QWEIGHTS  = {"flat": "0", "hf": "1", "adapt": "2"}

def run(cmd, **kw):
    return subprocess.run(cmd, capture_output=True, text=True, **kw)

def log(*a): print(*a, file=sys.stderr, flush=True)

# ---------------------------------------------------------------------------
# Build helpers
# ---------------------------------------------------------------------------
def build_c3d():
    bd = os.path.join(C3D, "build")
    lib = os.path.join(bd, "libc3d.a")
    if os.path.exists(lib):
        return lib
    r = run(["cmake", "-S", C3D, "-B", bd, "-DCMAKE_BUILD_TYPE=Release"])
    r = run(["cmake", "--build", bd, "--target", "c3d"])
    return lib if os.path.exists(lib) else None

def build_c4d():
    bd = os.path.join(C4D, "build")
    cli = os.path.join(bd, "c4d")
    if os.path.exists(cli):
        return cli
    args = ["cmake", "-S", C4D, "-B", bd, "-DCMAKE_BUILD_TYPE=Release"]
    if shutil.which("ninja"):
        args += ["-G", "Ninja"]
    run(args)
    run(["cmake", "--build", bd, "--target", "c4d_cli"])
    return cli if os.path.exists(cli) else None

def build_our_config(entropy, qwname, qwval, c3d_lib):
    """Compile compare_c3d_c4d.c against our codec sources for one config."""
    tag = f"{entropy}-{qwname}"
    outdir = os.path.join(REFB, tag)
    os.makedirs(outdir, exist_ok=True)
    binp = os.path.join(outdir, "cmp")
    srcs = [
        "src/codec.c", "src/transform/dct_int.c",
        "src/entropy/rice.c", "src/entropy/rlgr.c", "src/entropy/rans_static.c",
        "src/quant/subband.c", "src/ratectrl/fixed.c", "src/metrics/metrics.c",
        "harness/compare_c3d_c4d.c",
    ]
    srcs = [os.path.join(ROOT, s) for s in srcs]
    ent_map = {"rice": ("vc_rice_encode", "vc_rice_decode", "1"),
               "rlgr": ("vc_rlgr_encode", "vc_rlgr_decode", "2"),
               "rans": ("vc_rans_encode", "vc_rans_decode", "3")}
    enc, dec, etag = ent_map[entropy]
    cmd = ["cc", "-O3", "-ffp-contract=fast", "-ffast-math", "-std=c23",
           "-I", os.path.join(ROOT, "include"),
           f"-DVC_CHUNK_SIDE=64", f"-DVC_QWEIGHT={qwval}",
           f"-DVC_ENTROPY_ENC={enc}", f"-DVC_ENTROPY_DEC={dec}",
           f"-DVC_CFG_ENTROPY_TAG={etag}",
           f'-DVC_TAG_STR="{tag}"']
    if c3d_lib:
        cmd += ["-DHAVE_C3D", "-I", C3D]
    cmd += srcs
    if c3d_lib:
        cmd += [c3d_lib, "-fopenmp"]   # c3d's codec uses OpenMP internally
    cmd += ["-lm", "-o", binp]
    r = run(cmd)
    if r.returncode != 0:
        log(f"[build fail {tag}]\n{r.stderr[-2000:]}")
        return None
    return binp

# ---------------------------------------------------------------------------
# Inputs: assemble real PHerc Paris 4 volumes from the cached chunks.
# ---------------------------------------------------------------------------
def assemble_hires_256():
    """8 hires 128^3 chunks -> one 256^3 cube (2x2x2), saved to a temp raw."""
    d = os.path.join(DATA, "cache_hires")
    files = sorted(f for f in os.listdir(d) if f.endswith(".u8"))
    if len(files) < 8:
        return None
    C = 128
    vol = bytearray(256 * 256 * 256)
    order = files[:8]
    for idx, fn in enumerate(order):
        cz, cy, cx = (idx >> 2) & 1, (idx >> 1) & 1, idx & 1
        with open(os.path.join(d, fn), "rb") as fh:
            chunk = fh.read()
        for z in range(C):
            for y in range(C):
                src = ((z * C) + y) * C
                vz, vy, vx = cz*C+z, cy*C+y, cx*C
                dst = ((vz*256)+vy)*256 + vx
                vol[dst:dst+C] = chunk[src:src+C]
    p = os.path.join(REFB, "hires_256.u8")
    with open(p, "wb") as fh: fh.write(vol)
    return (p, 256, 256, 256, "hires-256")

def assemble_coarse_256():
    """Crop the coarse 384^3 volume to a 256^3 sub-region."""
    src = os.path.join(DATA, "coarse", "coarse_z15_y8_x8_3x3x3.u8")
    if not os.path.exists(src):
        return None
    S = 384
    with open(src, "rb") as fh: raw = fh.read()
    if len(raw) < S*S*S:
        return None
    C = 256
    vol = bytearray(C*C*C)
    for z in range(C):
        for y in range(C):
            s = ((z*S)+y)*S
            d = ((z*C)+y)*C
            vol[d:d+C] = raw[s:s+C]
    p = os.path.join(REFB, "coarse_256.u8")
    with open(p, "wb") as fh: fh.write(vol)
    return (p, 256, 256, 256, "coarse-256")

# ---------------------------------------------------------------------------
# c4d via CLI
# ---------------------------------------------------------------------------
def build_scorer():
    """Tiny binary: score <ref.u8> <rec.u8> <d> <h> <w> -> psnr ms_ssim gmsd."""
    binp = os.path.join(REFB, "score")
    src = os.path.join(REFB, "score.c")
    with open(src, "w") as f:
        f.write('''#include "../../include/vc/vc.h"
#include "../../src/metrics/metrics.h"
#include <stdio.h>
#include <stdlib.h>
static unsigned char* rd(const char*p,size_t*n){FILE*f=fopen(p,"rb");if(!f)return 0;
fseek(f,0,SEEK_END);long s=ftell(f);fseek(f,0,SEEK_SET);unsigned char*b=malloc(s);
fread(b,1,s,f);fclose(f);*n=s;return b;}
int main(int c,char**v){if(c<6)return 2;size_t a,b;
unsigned char*r=rd(v[1],&a),*q=rd(v[2],&b);
unsigned d=atoi(v[3]),h=atoi(v[4]),w=atoi(v[5]);
vc_metrics m;vc_compute_metrics(r,q,d,h,w,&m);
double g=vc_gmsd(r,q,d,h,w);
printf("%.2f %.4f %.4f\\n",m.psnr,m.ms_ssim,g);return 0;}''')
    r = run(["cc","-O2","-std=c23","-I",os.path.join(ROOT,"include"),
             src, os.path.join(ROOT,"src/metrics/metrics.c"),"-lm","-o",binp])
    if r.returncode != 0:
        log("[scorer build fail]\n"+r.stderr); return None
    return binp

def run_c4d(cli, scorer, raw, d, h, w, label, qs, rows):
    raw_sz = os.path.getsize(raw)
    for q in qs:
        arc = os.path.join(REFB, "c4d.tmp.c4d")
        out = os.path.join(REFB, "c4d.tmp.u8")
        import time
        t0 = time.time()
        e = run([cli, "encode", raw, arc, "--shape", f"{d},{h},{w}", "--q", str(q)])
        t1 = time.time()
        if e.returncode != 0 or not os.path.exists(arc):
            log(f"[c4d encode q={q} fail] {e.stderr[:200]}"); continue
        dd = run([cli, "decode", arc, out]); t2 = time.time()
        if dd.returncode != 0 or not os.path.exists(out):
            log(f"[c4d decode q={q} fail] {dd.stderr[:200]}"); continue
        alen = os.path.getsize(arc)
        sc = run([scorer, raw, out, str(d), str(h), str(w)])
        if sc.returncode != 0: log("[score fail]"); continue
        psnr, ssim, gmsd = sc.stdout.split()
        ratio = raw_sz / alen
        enc_mbs = raw_sz/1e6/max(t1-t0,1e-9); dec_mbs = raw_sz/1e6/max(t2-t1,1e-9)
        rows.append(("c4d", label, float(q), ratio, float(psnr), float(ssim),
                     float(gmsd), enc_mbs, dec_mbs, "ref-c4d"))

# ---------------------------------------------------------------------------
def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--inputs", default="both", choices=["hires","coarse","both"])
    ap.add_argument("--q", nargs="*", type=float,
                    default=[2,4,8,16,32,64,128])
    ap.add_argument("--configs", default="all",
                    help="comma list of entropy-qweight tags, or 'all' or 'lead'")
    args = ap.parse_args()
    os.makedirs(REFB, exist_ok=True)

    log("== building references ==")
    c3d_lib = build_c3d()
    log(f"c3d: {'OK '+c3d_lib if c3d_lib else 'FAILED (continuing without)'}")
    c4d_cli = build_c4d()
    log(f"c4d: {'OK '+c4d_cli if c4d_cli else 'FAILED (continuing without)'}")
    scorer = build_scorer()

    # select our configs
    if args.configs == "all":
        combos = [(e,wn,wv) for e in ENTROPIES for wn,wv in QWEIGHTS.items()]
    elif args.configs == "lead":
        combos = [("rlgr","hf",QWEIGHTS["hf"]), ("rans","hf",QWEIGHTS["hf"]),
                  ("rice","flat",QWEIGHTS["flat"])]
    else:
        combos = []
        for tag in args.configs.split(","):
            e, wn = tag.split("-"); combos.append((e, wn, QWEIGHTS[wn]))

    log("== building our %d configs ==" % len(combos))
    ourbins = {}
    for e,wn,wv in combos:
        b = build_our_config(e, wn, wv, c3d_lib)
        if b: ourbins[f"{e}-{wn}"] = b

    inputs = []
    if args.inputs in ("hires","both"):
        x = assemble_hires_256();  inputs.append(x) if x else log("[no hires data]")
    if args.inputs in ("coarse","both"):
        x = assemble_coarse_256(); inputs.append(x) if x else log("[no coarse data]")

    rows = []
    qstr = [str(q) for q in args.q]
    for raw,d,h,w,label in inputs:
        log(f"== input {label} ({d}x{h}x{w}) ==")
        # our configs + c3d (c3d runs inside the FIRST our-binary that has it
        # linked; to avoid duplicate c3d rows, run c3d only once via any binary)
        c3d_done = False
        for tag, binp in ourbins.items():
            r = run([binp, raw, str(d), str(h), str(w), label] + qstr)
            if r.returncode != 0:
                log(f"[run {tag} fail] {r.stderr[:300]}"); continue
            for line in r.stdout.strip().splitlines():
                f = line.split("\t")
                codec = f[0]
                if codec == "c3d":
                    if c3d_done: continue
                    rows.append((codec,f[1],float(f[2]),float(f[3]),float(f[4]),
                                 float(f[5]),float(f[6]),float(f[7]),float(f[8]),f[9]))
                else:
                    rows.append((codec,f[1],float(f[2]),float(f[3]),float(f[4]),
                                 float(f[5]),float(f[6]),float(f[7]),float(f[8]),tag))
            c3d_done = True
        # c4d
        if c4d_cli and scorer:
            run_c4d(c4d_cli, scorer, raw, d, h, w, label, args.q, rows)

    # ---- print the side-by-side table ----
    print("\n# codec/config\tinput\tq\tratio\tPSNR\tMS-SSIM\tGMSD\tenc MB/s\tdec MB/s")
    rows.sort(key=lambda r:(r[1], -r[3]))
    for r in rows:
        print(f"{r[9]:<16}\t{r[1]:<11}\t{r[2]:>6.4g}\t{r[3]:>7.1f}\t{r[4]:>6.2f}"
              f"\t{r[5]:>6.4f}\t{r[6]:>6.4f}\t{r[7]:>7.0f}\t{r[8]:>7.0f}")

if __name__ == "__main__":
    main()
